llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -8287,6 +8289,13 @@ AArch64TargetLowering::LowerCall(CallLoweringInfo &CLI,
 
   SDValue InGlue;
   if (RequiresSMChange) {
+
+    if (Subtarget->hasSVE()) {
```

Comment:
I'm not sure I follow what the issue is... I think we discussed before that we want debuggers to be able to unwind the stack even if a function is nounwind.

Reply:
> when I compile some code with -fno-asynchronous-unwind-tables, it will still generate these directives. Perhaps this should have a diagnostic in Clang?

I'm happy to add such a diagnostic, but I want to make sure I have the reason for this requirement correct.
If I understand correctly, enabling asychronous unwind tables is required now that I am emitting the save and restore of VG at the point of each call in the function which changes streaming-mode.

> we want debuggers to be able to unwind the stack even if a function is nounwind.

Given this, I'm not entirely sure why we would expect to be able to unwind if `-fno-asynchronous-unwind-tables` was used or if the function has been marked as nounwind? I might have misunderstood something though!

Reply:
There are two different kinds of DWARF "unwind info"; one is the kind that's in a loadable section, and used for EH. The other is in a debug info section, not loaded at runtime.  If you specify `-fasynchronous-unwind-tables`, you get the former; if you specify `-g -fno-asynchronous-unwind-tables`, you get the latter.

If you request no debug info and no unwind tables, we shouldn't emit any DWARF directives.

But like I mentioned, I think we want to unconditionally emit the code to save the VG, whether or not we emit the corresponding DWARF directives.

Reply:
If we still emit the spill of VG with `-fno-asynchronous-unwind-tables`, I don't know how we would be able to recover the correct value without the per-call CFI saves and restores. And without the correct VG value, I don't think it will be possible to recover any VG based values in the stack frame.

I'm still not sure why this wouldn't require a diagnostic, because if there is not enough information to recover VG then I don't think we can unwind correctly?

Reply:
If there's no DWARF unwind, nothing can unwind the stack whether or not we store the VG.

Reply:
@efriedma-quic I think the issue is that when we emit unwind info that is _not_ asynchronous, then the unwinder can't correctly unwind the stack because it would use the wrong value for VG to compute the offsets of callee-saves. So any unwind info that would be produced is broken.

Reply:
Oh, right, there's also that dimension.  I'm not sure I understand the interaction here, but if there's an issue, can we just force on "asynchronous" unwind info in that case?  The point of non-async unwind info isn't that it's a different unwind format; it's just an optimization to reduce the size of the unwind info.

Reply:
I've made changes to `needsAsyncDwarfUnwindInfo` in AArch64MachineFunctionInfo.cpp to always return true if the function has streaming-mode changes. I believe this will ensure we emit the correct information when `-fno-asynchronous-unwind-tables` is specified.